--- XFree86-4.0/xc/programs/Xserver/cfb/Imakefile.asmflags	Thu Jun  1 11:52:12 2000
+++ XFree86-4.0/xc/programs/Xserver/cfb/Imakefile	Thu Jun  1 11:47:38 2000
@@ -234,10 +234,10 @@
 	$(RM) stipsparcte.s
 
 stipsparc.o: stipsparc.s
-	$(CPP) stipsparc.s | $(AS) -o $@ -
+	$(CPP) stipsparc.s AsmDefines | $(AS) -o $@ -
 
 stipsparcte.o: stipsparcte.s
-	$(CPP) -DTETEXT stipsparcte.s | $(AS) -o $@ -
+	$(CPP) -DTETEXT stipsparcte.s AsmDefines | $(AS) -o $@ -
 #endif
 #endif
 
@@ -250,10 +250,10 @@
 	$(RM) stipsprcte32.s
 
 stipsprc32.o: stipsprc32.s
-	$(CPP) stipsprc32.s | $(AS) -o $@ -
+	$(CPP) stipsprc32.s AsmDefines | $(AS) -o $@ -
 
 stipsprcte32.o: stipsprcte32.s
-	$(CPP) -DTETEXT stipsprcte32.s | $(AS) -o $@ -
+	$(CPP) -DTETEXT stipsprcte32.s AsmDefines | $(AS) -o $@ -
 #endif
 #endif
 
