--- XFree86-3.3.6/xc/programs/Xserver/Imakefile.parallel	Thu Jan 20 12:52:48 2000
+++ XFree86-3.3.6/xc/programs/Xserver/Imakefile	Thu Jan 20 12:59:14 2000
@@ -803,6 +803,10 @@ XF86S3SYSLIBS = $(SYSLIBS)
 #if HasParallelMake
 MakeMutex($(XF86S3SUBDIRS) $(XF86S3OBJS) $(XF86S3LIBS) $(XF86S3SYSLIBS))
 #endif
+#if HasGnuMake
+$(XF86S3OBJS) $(XF86S3LIBS) $(XF86S3SYSLIBS):: $(XF86S3SUBDIRS)
+	@if [ -f $@ ]; then touch $@; fi
+#endif
 SetUIDServerTarget(XF86_S3,$(XF86S3SUBDIRS),$(XF86S3OBJS),$(XF86S3LIBS),$(XF86S3SYSLIBS))
 #ifndef ServerToInstall
 #define ServerToInstall XF86_S3
@@ -1802,6 +1806,10 @@ CFB16DIR = cfb16
 CFB32DIR = cfb32
 #if HasParallelMake
 MakeMutex($(STDDIRS) $(MFBDIR) $(CFB8DIR) $(CFB16DIR) $(CFB32DIR) $(DEPDIRS))
+#endif
+#if HasGnuMake
+$(MFBDIR) $(CFB8DIR) $(CFB16DIR) $(CFB32DIR) $(DEPDIRS):: $(STDDIRS)
+	@if [ -f $@ ]; then touch $@; fi
 #endif
 Xnon: $(STDDIRS) $(MFBDIR) $(CFB8DIR) $(CFB16DIR) $(CFB32DIR) $(DEPDIRS)
 #endif /* XnonServer */
--- XFree86-3.3.6/xc/Makefile.parallel	Wed Jun 23 23:00:33 1999
+++ XFree86-3.3.6/xc/Makefile	Thu Jan 20 08:58:18 2000
@@ -88,7 +88,7 @@ $(DEPENDSRC)/Makefile.proto: imake.proto
 
 depend.bootstrap: $(DEPENDSRC)/Makefile.proto
 	cd $(DEPENDSRC) && $(RM) -r Makefile Makefile.dep makedepend *.o bootstrap
-	cd $(DEPENDSRC) && $(MAKE) -f Makefile.proto bootstrap
+	cd $(DEPENDSRC) && $(MAKE) -f Makefile.proto makedepend
 
 $(IMAKESRC)/Makefile.proto: depend.bootstrap
 	$(IMAKE_CMD) -s $(IMAKESRC)/Makefile.proto -f $(IMAKESRC)/Imakefile -DTOPDIR=$(IMAKETOP) -DCURDIR=$(IMAKESRC)
