--- XFree86-4.0/xc/programs/Xserver/os/secauth.c.foo	Tue May 16 19:29:26 2000
+++ XFree86-4.0/xc/programs/Xserver/os/secauth.c	Tue May 16 19:33:07 2000
@@ -42,10 +42,10 @@
     ClientPtr	client,
     char	**reason)
 {
-    char	*policy = *dataP;
+    unsigned char	*policy = *dataP;
     int		length;
     Bool	permit;
-    int		nPolicies;
+    unsigned int	nPolicies;
     char	**sitePolicies;
     int		nSitePolicies;
     Bool	found = FALSE;
@@ -63,7 +63,8 @@
     sitePolicies = SecurityGetSitePolicyStrings(&nSitePolicies);
 
     while (nPolicies) {
-	int strLen, sitePolicy;
+	unsigned int strLen;
+	int sitePolicy;
 
 	if (length == 0) {
 	    *reason = InvalidPolicyReason;
