--- xc/programs/Xserver/hw/xfree68/fbdev/fbdev.c.jj	Wed Jun  2 07:40:47 1999
+++ xc/programs/Xserver/hw/xfree68/fbdev/fbdev.c	Mon Jun  5 17:13:55 2000
@@ -27,6 +27,8 @@
 
 #define DIRECTCOLORHACK		/* hack for directcolor */
 
+#include <stdlib.h>
+#include <math.h>
 #include "X.h"
 #include "input.h"
 #include "scrnintstr.h"
@@ -51,8 +53,6 @@
 #endif /* CONFIG_AFB */
 #include <linux/fb.h>
 #include <asm/page.h>
-#include <stdlib.h>
-#include <math.h>
 
 #include "colormapst.h"
 #include "resource.h"
