--- XFree86-3.3.6/xc/config/cf/xfree86.cf.svgaprobe	Tue Feb 29 13:09:36 2000
+++ XFree86-3.3.6/xc/config/cf/xfree86.cf	Tue Feb 29 13:36:16 2000
@@ -433,9 +433,9 @@
 #  define XF86SvgaDrivers	mga tvga8900 s3v
 # else
 #  if defined(LinuxArchitecture)
-#   define XF86SvgaDrivers	nv et4000 et3000 pvga1 gvga r128 ati sis tvga8900 \
+#   define XF86SvgaDrivers	s3_savage nv et4000 et3000 pvga1 gvga r128 ati sis tvga8900 \
 				cirrus ncr77c22 compaq mga oak al2101 \
-				ali cl64xx video7 ark mx realtek s3_savage s3v apm \
+				ali cl64xx video7 ark mx realtek s3v apm \
 				s3_svga neo chips cyrix rendition p9x00 \
 				spc8110 i740 tdfx i810 smi generic
 #  else
